Source/Core/Core/HW/DVD/AMMediaboard.cpp

```diff
@@ -0,0 +1,1855 @@
+  if (addr->sin_addr.s_addr == inet_addr("192.168.13.1"))
+  {
+    addr->sin_addr.s_addr = inet_addr("10.0.0.45");
+  }
+
+  addr->sin_family = Common::swap16(addr->sin_family);
+  //*(u32*)(&addr.sin_addr) = Common::swap32(*(u32*)(&addr.sin_addr));
+
+  int ret = 0;
+  int err = 0;
+  u_long val = 1;
+
+  // Set socket to non-blocking
+  ioctlsocket(fd, FIONBIO, &val);
+
+  ret = connect(fd, (const sockaddr*)addr, len);
+  err = WSAGetLastError();
+
+  if (ret == SOCKET_ERROR && err == WSAEWOULDBLOCK)
+  {
+    fd_set writefds;
+    FD_ZERO(&writefds);
+    FD_SET(fd, &writefds);
+
+    timeval timeout;
+    timeout.tv_sec = 0;
+    timeout.tv_usec = s_timeouts[0];
+
+    ret = select(0, NULL, &writefds, NULL, &timeout);
+    if (ret > 0 && FD_ISSET(fd, &writefds))
+    {
+      int so_error = 0;
+      socklen_t optlen = sizeof(so_error);
+      if (getsockopt(fd, SOL_SOCKET, SO_ERROR, (char*)&so_error, &optlen) == 0 && so_error == 0)
+      {
+        s_last_error = SSC_SUCCESS;
+        ret = 0;
+      }
+      else
+      {
+        s_last_error = SOCKET_ERROR;
+        ret = SOCKET_ERROR;
+      }
+    }
+    else if (ret == 0)
+    {
+      // Timeout
+      s_last_error = SSC_EWOULDBLOCK;
+      ret = SOCKET_ERROR;
+    }
+    else
+    {
+      // select() failed
+      s_last_error = SOCKET_ERROR;
+      ret = SOCKET_ERROR;
+    }
+  }
+  else if (ret == SOCKET_ERROR)
+  {
+    // Immediate failure (e.g. WSAECONNREFUSED)
+    s_last_error = ret;
+  }
+  else
+  {
+    s_last_error = SSC_SUCCESS;
+  }
+
+  // Restore blocking mode
+  val = 0;
+  ioctlsocket(fd, FIONBIO, &val);
+
+  return ret;
+}
+static void FileWriteData(File::IOFile* file, u32 seek_pos, const u8* data, size_t length)
+{
+  file->Seek(seek_pos, File::SeekOrigin::Begin);
+  file->WriteBytes(data, length);
+  file->Flush();
+}
+u32 ExecuteCommand(std::array<u32, 3>& DICMDBUF, u32 address, u32 length)
+{
+  auto& system = Core::System::GetInstance();
+  auto& memory = system.GetMemory();
+
+  DICMDBUF[0] ^= s_GCAM_key_a;
+  DICMDBUF[1] ^= s_GCAM_key_b;
+  DICMDBUF[2] ^= s_GCAM_key_c;
+
+  u32 seed = DICMDBUF[0] >> 16;
+
+  s_GCAM_key_a *= seed;
+  s_GCAM_key_b *= seed;
+  s_GCAM_key_c *= seed;
+
+  /*
+     Key setup for Triforce IPL:
+     These RAM offset always hold the keys for the next command and since it sends two dummy
+     commands before a real read we can just use the key from RAM without missing any real commands.
+  */
+  if (s_GCAM_key_a == 0)
+  {
+    if (memory.Read_U32(0))
+    {
+      HLE::Patch(system, 0x813048B8, "OSReport");
+      HLE::Patch(system, 0x8130095C, "OSReport");  // Apploader
+
+      InitKeys(memory.Read_U32(0), memory.Read_U32(4), memory.Read_U32(8));
+    }
+  }
+
+  DICMDBUF[0] <<= 24;
+  DICMDBUF[1] <<= 2;
+
+  // SegaBoot adds bits for some reason to offset/length
+  // also adds 0x20 to offset
+  if (DICMDBUF[1] == 0x00100440)
+  {
+    s_segaboot = true;
+  }
+
+  u32 command = DICMDBUF[0];
+  u32 offset = DICMDBUF[1];
```

Comment:
Since `DICMDBUF` is passed by reference, this also modifies the values in DVDInterface (which I guess is why the `m_DICMDBUF[1] >>= 2;` line exists when falling through to normal DI logic). It makes sense to apply the decryption, but I feel like these shifts make less sense. It would be better to change the lines below to use

```suggestion
  u32 command = DICMDBUF[0] << 24;
  u32 offset = DICMDBUF[1] << 2;

  // SegaBoot adds bits for some reason to offset/length
  // also adds 0x20 to offset
  if (offset == 0x00100440)
  {
    s_segaboot = true;
  }
```

I assume the `<< 24` part is to address this (from https://debugmo.de/2008/05/part-2-dumping-the-media-board/):

> However, DI commands are still different to the original one. For example, the “READ” command is usually a8000000, whereas here it’s “000100a8”.